src/ClassyResource.js

```diff
@@ -75,46 +73,106 @@ export default class ClassyResource {
+
+        if (!response) {
+          return this._makeRequest(requestPath, requestMethod, requestHeaders, form, data);
+        }
+
+        if (response === 'app') {
+          return this._refreshAppToken().then((response) => {
+            const appToken = _.get(this._classy.appToken, 'access_token', null);
+            delete data.token;
+
+            requestHeaders.Authorization = 'Bearer ' + appToken;
+
+            return this._makeRequest(requestPath, requestMethod, requestHeaders, form, data);
+          }, (error) => {
+            console.error('App token refresh failed: ', error);
+          });
+        }
+
+        if (response === 'member') {
+          const memberToken = _.get(data, 'token.access_token', false);
+          delete data.token;
+
+          requestHeaders.Authorization = 'Bearer ' + memberToken;
+
+          return this._makeRequest(requestPath, requestMethod, requestHeaders, form, data);
+        }
+
+      }, (error) => {
+        throw new Error('No token defined. Expected memberToken object or `token: \'app\'`');
       });
 
-      /** Merge default headers with spec headers */
-      const DEFAULT_REQUEST_HEADERS = {
-        Authorization: 'Bearer ' + token.value,
-        Accept: 'application/json',
-        'Content-Type': 'application/json'
-      };
+    };
+  }
 
-      const requestHeaders = this._classy.headers
-        || _.merge(DEFAULT_REQUEST_HEADERS, spec.headers);
+  /**
+   *
+   */
+  _chooseToken(form, data) {
+    const dataToken = _.get(data, 'token', false);
+    const clientCredentialsRequest = _.get(form, 'grant_type', false) === 'client_credentials';
+
+    const promise = new Promise((resolve, reject) => {
+      /** Unless it's a `client_credentials` request, pick a token */
+      if (clientCredentialsRequest) {
+        resolve(false);
+      } else {
+        if (!dataToken) {
+          reject('No token defined. Expected memberToken object or `token: \'app\'`');
+        } else if (dataToken === 'app') {
+          resolve('app');
+        } else {
+          resolve('member');
+        }
 
-      /** Handle auth requests */
-      let form = false;
-      if (isAuthRequest) {
-        form = this._generateAuthForm(args);
       }
+    });
 
-      /** Make the request and return a promise */
-      return this._makeRequest(requestPath, requestMethod, requestHeaders, form, data);
-    };
+    return promise;
+  }
+
+  /**
+   * If this._classy.appToken exists, just resolve,
+   * if not, refresh and then resolve.
+   * @return {[type]} [description]
+   */
+  _refreshAppToken() {
+    const promise = new Promise((resolve, reject) => {
+      if (!this._classy.appToken) {
+        this._classy.getAppToken().then((response) => {
+          this._classy.setAppToken(response);
+          resolve(response);
+        }, (error) => {
+          reject(error);
+        });
+      } else {
+        resolve(this._classy.appToken);
+      }
+    });
```

Comment:
It seems to me you could still have concurrency issues here if an app token expires and then multiple requests come in simultaneously, each fetch a new app token, then overwrite each other with the result, causing all but one to fail. Might want to have a global "pendingAppToken" flag/promise that you can check and/or .then() off of if appToken is null.